### PR TITLE
enhance(datastore): dump wal to the local disk before sending to the remote storage

### PR DESCRIPTION
**Pull request by xuchuan**
## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [ ] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                var entry = entries.get(i);
+                var entrySize = CodedOutputStream.computeMessageSizeNoTag(entry);
+                if (WalManager.HEADER.length + entrySize > this.walMaxFileSize) {
+                    log.error("data loss: discard unexpected huge entry. "
+                                    + "size={} table={} schema={} records count={}",
+                            entry.getSerializedSize(),
+                            entry.getTableName(),
+                            entry.getTableSchema(),
+                            entry.getRecordsCount());
+                }
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java:141]: There will be a dead loop if the first entry is too large without `continue`

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        if (this.outputStream != null) {
+            try {
+                this.outputStream.close();
+            } catch (IOException e) {
+                log.error("failed to close", e);
+            }
+        }
+        this.currentFileSize = WalManager.HEADER.length;
+        var info = new LogFileInfo(this.logFileIndex);
+        this.logFileMap.put(this.logFileIndex, info);
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java:214]: The `logFileMap` will increase without limit

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalRemoteFileManager.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                    var entry = this.logFileToPush.firstEntry();
+                    info = entry.getValue();
+                }
+            }
+            try {
+                try (var in = Files.newInputStream(info.getPath(), StandardOpenOption.READ)) {
+                    var b = in.readNBytes(info.getSize());
+                    if (b.length != info.getSize()) {
+                        throw new RuntimeException(
+                                MessageFormat.format("invalid file {}, expected size {0}, actual {1}",
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalRemoteFileManager.java:166]: Missing offset in placeholder?

**Review by jialeicui — APPROVED** [review]

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalRemoteFileManager.java (hunk 3 above)
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalRemoteFileManager.java]: Done

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java (hunk 2 above)
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java]: Done

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java (hunk 1 above)
Comment (xuchuan) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java]: Done

**Review by jialeicui — APPROVED** [review]
LGTM

**Pull request merged**